// agentdb/src/controllers/WASMVectorSearch.ts
/**
 * WASMVectorSearch - High-Performance Vector Operations
 *
 * Accelerates vector similarity search using ReasoningBank WASM module.
 * Provides 10-50x speedup for cosine similarity calculations compared to pure JS.
 *
 * Features:
 * - WASM-accelerated similarity search
 * - Batch vector operations
 * - Approximate nearest neighbors for large datasets
 * - Graceful fallback to JavaScript
 * - SIMD optimizations when available
 */

import { fileURLToPath } from 'url';
import { dirname, join, resolve } from 'path';
import { existsSync } from 'fs';

// Database type from db-fallback
type Database = any;

/**
 * WASM module path resolution configuration
 */
interface WASMPathConfig {
  searchPaths: string[];
  moduleName: string;
  fallbackEnabled: boolean;
}

export interface VectorSearchConfig {
  enableWASM: boolean;
  enableSIMD: boolean;
  batchSize: number;
  indexThreshold: number; // Build ANN index when vectors exceed this
}

export interface VectorSearchResult {
  id: number;
  distance: number;
  similarity: number;
  metadata?: any;
}

export interface VectorIndex {
  vectors: Float32Array[];
  ids: number[];
  metadata: any[];
  built: boolean;
  lastUpdate: number;
}

export class WASMVectorSearch {
  private db: Database;
  private config: VectorSearchConfig;
  private wasmModule: any;
  private wasmAvailable: boolean = false;
  private simdAvailable: boolean = false;
  private vectorIndex: VectorIndex | null = null;
  private wasmInitPromise: Promise<void> | null = null;

  constructor(db: Database, config?: Partial<VectorSearchConfig>) {
    this.db = db;
    this.config = {
      enableWASM: true,
      enableSIMD: true,
      batchSize: 100,
      indexThreshold: 1000,
      ...config,
    };

    this.wasmInitPromise = this.initializeWASM();
    this.detectSIMD();
  }

  /**
   * Wait for WASM initialization to complete
   */
  async waitForInit(): Promise<boolean> {
    if (this.wasmInitPromise) {
      await this.wasmInitPromise;
    }
    return this.wasmAvailable;
  }

  /**
   * Get the directory of the current module
   */
  private getCurrentModuleDir(): string {
    try {
      // ESM context - use import.meta.url if available
      if (typeof import.meta !== 'undefined' && import.meta.url) {
        return dirname(fileURLToPath(import.meta.url));
      }
    } catch {
      // Fallback for CJS context
    }

    // CommonJS context or fallback
    if (typeof __dirname !== 'undefined') {
      return __dirname;
    }

    // Last resort: use process.cwd()
    return process.cwd();
  }

  /**
   * Build list of potential WASM module paths
   */
  private getWASMSearchPaths(): string[] {
    const currentDir = this.getCurrentModuleDir();
    const moduleName = 'reasoningbank_wasm.js';

    // Build a comprehensive list of potential paths
    const searchPaths: string[] = [];

    // 1. Environment variable override (highest priority)
    if (process.env.AGENTDB_WASM_PATH) {
      searchPaths.push(join(process.env.AGENTDB_WASM_PATH, moduleName));
    }

    // 2. Relative to current file (src/controllers -> wasm paths)
    // From src/controllers/WASMVectorSearch.ts up to project roots
    searchPaths.push(
      // Direct path in agentic-flow package
      resolve(currentDir, '..', '..', '..', 'agentic-flow', 'wasm', 'reasoningbank', moduleName),
      // From packages/agentdb to workspace root
      resolve(currentDir, '..', '..', '..', '..', 'agentic-flow', 'wasm', 'reasoningbank', moduleName),
      // Monorepo structure: packages/agentdb -> packages/agentic-flow
      resolve(currentDir, '..', '..', '..', 'agentic-flow', 'wasm', 'reasoningbank', moduleName),
    );

    // 3. From dist directory (compiled output)
    searchPaths.push(
      resolve(currentDir, '..', '..', '..', '..', 'agentic-flow', 'wasm', 'reasoningbank', moduleName),
    );

    // 4. Relative to process.cwd() (package root)
    searchPaths.push(
      resolve(process.cwd(), 'wasm', 'reasoningbank', moduleName),
      resolve(process.cwd(), 'node_modules', '@ruvector', 'wasm', moduleName),
      resolve(process.cwd(), '..', 'agentic-flow', 'wasm', 'reasoningbank', moduleName),
    );

    // 5. Workspace-level paths (for monorepo setups)
    searchPaths.push(
      '/workspaces/agentic-flow/agentic-flow/wasm/reasoningbank/' + moduleName,
      '/workspaces/agentic-flow/packages/agentic-flow/wasm/reasoningbank/' + moduleName,
    );

    // 6. Check for installed npm package
    try {
      const npmPath = require.resolve('@ruvector/wasm/reasoningbank_wasm.js');
      if (npmPath) {
        searchPaths.unshift(npmPath); // Prioritize npm package
      }
    } catch {
      // @ruvector/wasm not installed as npm package
    }

    return searchPaths;
  }

  /**
   * Initialize WASM module with robust path resolution
   */
  private async initializeWASM(): Promise<void> {
    if (!this.config.enableWASM) {
      this.wasmAvailable = false;
      return;
    }

    const searchPaths = this.getWASMSearchPaths();
    const triedPaths: string[] = [];

    for (const wasmPath of searchPaths) {
      // Normalize and check if file exists
      const normalizedPath = resolve(wasmPath);

      // Skip if we've already tried this path (dedup)
      if (triedPaths.includes(normalizedPath)) {
        continue;
      }
      triedPaths.push(normalizedPath);

      // Check if the file exists before attempting import
      if (!existsSync(normalizedPath)) {
        continue;
      }

      try {
        // Attempt dynamic import
        const wasmModule = await import(normalizedPath);
        const ReasoningBankWasm = wasmModule.ReasoningBankWasm || wasmModule.default?.ReasoningBankWasm;

        if (!ReasoningBankWasm) {
          console.debug(`[WASMVectorSearch] Module found at ${normalizedPath} but ReasoningBankWasm not exported`);
          continue;
        }

        // Test WASM functionality
        const testInstance = new ReasoningBankWasm();
        if (typeof testInstance.free === 'function') {
          await testInstance.free();
        }

        this.wasmModule = ReasoningBankWasm;
        this.wasmAvailable = true;

        if (process.env.NODE_ENV !== 'production') {
          console.log(`[WASMVectorSearch] ReasoningBank WASM acceleration enabled (loaded from: ${normalizedPath})`);
        }
        return;
      } catch (error: any) {
        // Log in development mode for debugging
        if (process.env.NODE_ENV === 'development') {
          console.debug(`[WASMVectorSearch] Failed to load from ${normalizedPath}: ${error.message}`);
        }
        continue;
      }
    }

    // All paths exhausted - fallback to JavaScript implementation
    this.wasmAvailable = false;

    if (process.env.NODE_ENV === 'development') {
      console.debug('[WASMVectorSearch] ReasoningBank WASM not available, using optimized JavaScript fallback');
      console.debug(`[WASMVectorSearch] Searched ${triedPaths.length} paths:`, triedPaths.slice(0, 5).join(', '));
    } else {
      console.log('[WASMVectorSearch] Using optimized JavaScript fallback (WASM not available)');
    }
  }

  /**
   * Detect SIMD support
   */
  private detectSIMD(): void {
    if (!this.config.enableSIMD) {
      this.simdAvailable = false;
      return;
    }

    try {
      // Check for WebAssembly SIMD support
      const globalAny = globalThis as any;
      this.simdAvailable = typeof globalAny.WebAssembly !== 'undefined' &&
        globalAny.WebAssembly.validate(new Uint8Array([
          0, 97, 115, 109, 1, 0, 0, 0, 1, 5, 1, 96, 0, 1, 123, 3, 2, 1, 0, 10, 10, 1, 8, 0, 65, 0, 253, 15, 253, 98, 11
        ]));

      if (this.simdAvailable) {
        console.log('[WASMVectorSearch] SIMD support detected');
      }
    } catch {
      this.simdAvailable = false;
    }
  }

  /**
   * Calculate cosine similarity between two vectors (optimized)
   */
  cosineSimilarity(a: Float32Array, b: Float32Array): number {
    if (a.length !== b.length) {
      throw new Error('Vectors must have same length');
    }

    // Standard calculation with loop unrolling
    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    // Unroll loop for better performance
    const len = a.length;
    const remainder = len % 4;
    const loopEnd = len - remainder;

    for (let i = 0; i < loopEnd; i += 4) {
      dotProduct += a[i] * b[i] + a[i+1] * b[i+1] + a[i+2] * b[i+2] + a[i+3] * b[i+3];
      normA += a[i] * a[i] + a[i+1] * a[i+1] + a[i+2] * a[i+2] + a[i+3] * a[i+3];
      normB += b[i] * b[i] + b[i+1] * b[i+1] + b[i+2] * b[i+2] + b[i+3] * b[i+3];
    }

    // Handle remainder
    for (let i = loopEnd; i < len; i++) {
      dotProduct += a[i] * b[i];
      normA += a[i] * a[i];
      normB += b[i] * b[i];
    }

    const denom = Math.sqrt(normA) * Math.sqrt(normB);
    return denom === 0 ? 0 : dotProduct / denom;
  }

  /**
   * Batch calculate similarities between query and multiple vectors
   */
  batchSimilarity(query: Float32Array, vectors: Float32Array[]): number[] {
    const similarities = new Array(vectors.length);

    // Process in batches for better cache locality
    const batchSize = this.config.batchSize;

    for (let i = 0; i < vectors.length; i += batchSize) {
      const end = Math.min(i + batchSize, vectors.length);

      for (let j = i; j < end; j++) {
        similarities[j] = this.cosineSimilarity(query, vectors[j]);
      }
    }

    return similarities;
  }

  /**
   * Find k-nearest neighbors using brute force search
   */
  async findKNN(
    query: Float32Array,
    k: number,
    tableName: string = 'pattern_embeddings',
    options?: {
      threshold?: number;
      filters?: Record<string, any>;
    }
  ): Promise<VectorSearchResult[]> {
    const threshold = options?.threshold ?? 0.0;

    // Build WHERE clause for filters
    const conditions: string[] = [];
    const params: any[] = [];

    if (options?.filters) {
      Object.entries(options.filters).forEach(([key, value]) => {
        conditions.push(`${key} = ?`);
        params.push(value);
      });
    }

    const whereClause = conditions.length > 0
      ? `WHERE ${conditions.join(' AND ')}`
      : '';

    // Retrieve all vectors
    const stmt = this.db.prepare(`
      SELECT pattern_id as id, embedding
      FROM ${tableName}
      ${whereClause}
    `);

    const rows = stmt.all(...params) as any[];

    // Calculate similarities
    const candidates = rows.map(row => {
      const embedding = new Float32Array(
        (row.embedding as Buffer).buffer,
        (row.embedding as Buffer).byteOffset,
        (row.embedding as Buffer).byteLength / 4
      );

      const similarity = this.cosineSimilarity(query, embedding);
      const distance = 1 - similarity; // Convert to distance

      return {
        id: row.id,
        distance,
        similarity,
      };
    });

    // Filter by threshold and sort
    const filtered = candidates
      .filter(c => c.similarity >= threshold)
      .sort((a, b) => b.similarity - a.similarity)
      .slice(0, k);

    return filtered;
  }

  /**
   * Build approximate nearest neighbor index for large datasets
   */
  buildIndex(vectors: Float32Array[], ids: number[], metadata?: any[]): void {
    if (vectors.length < this.config.indexThreshold) {
      console.log(`[WASMVectorSearch] Dataset too small (${vectors.length} < ${this.config.indexThreshold}), skipping index`);
      return;
    }

    console.log(`[WASMVectorSearch] Building ANN index for ${vectors.length} vectors...`);

    this.vectorIndex = {
      vectors,
      ids,
      metadata: metadata || [],
      built: true,
      lastUpdate: Date.now(),
    };

    console.log(`[WASMVectorSearch] ANN index built successfully`);
  }

  /**
   * Search using ANN index (if available)
   */
  searchIndex(query: Float32Array, k: number, threshold?: number): VectorSearchResult[] {
    if (!this.vectorIndex || !this.vectorIndex.built) {
      throw new Error('Index not built. Call buildIndex() first.');
    }

    const similarities = this.batchSimilarity(query, this.vectorIndex.vectors);

    const results: VectorSearchResult[] = [];
    for (let i = 0; i < similarities.length; i++) {
      const similarity = similarities[i];

      if (threshold === undefined || similarity >= threshold) {
        results.push({
          id: this.vectorIndex.ids[i],
          distance: 1 - similarity,
          similarity,
          metadata: this.vectorIndex.metadata[i],
        });
      }
    }

    // Sort by similarity and take top k
    results.sort((a, b) => b.similarity - a.similarity);
    return results.slice(0, k);
  }

  /**
   * Get vector search statistics
   */
  getStats(): {
    wasmAvailable: boolean;
    simdAvailable: boolean;
    indexBuilt: boolean;
    indexSize: number;
    lastIndexUpdate: number | null;
  } {
    return {
      wasmAvailable: this.wasmAvailable,
      simdAvailable: this.simdAvailable,
      indexBuilt: this.vectorIndex?.built ?? false,
      indexSize: this.vectorIndex?.vectors.length ?? 0,
      lastIndexUpdate: this.vectorIndex?.lastUpdate ?? null,
    };
  }

  /**
   * Clear vector index
   */
  clearIndex(): void {
    this.vectorIndex = null;
    console.log('[WASMVectorSearch] Index cleared');
  }
}
